File: src/services/documentService.ts
import {
  ACCEPTED_MIME_TYPES,
  DOCUMENT_TYPE_BY_ID,
  MAX_FILE_SIZE_BYTES,
} from "@/data"
import type {
  DocumentExtraction,
  DocumentTypeId,
  FileValidationResult,
  ProcessingStage,
  UploadedDocument,
} from "@/types"
import { formatFileSize } from "@/utils"
import { API_CONFIG, delay, request } from "./apiClient"

/**
 * The pipeline stages, in order. The mock walks through these on a timer; the
 * backend will report the same sequence over
 * FastAPI → PaddleOCR → vision/LLM extraction → PostgreSQL.
 */
export const PIPELINE_STAGES: ProcessingStage[] = [
  "uploading",
  "ocr",
  "extracting",
  "structuring",
  "complete",
]

const EXTENSION_FALLBACK: Record<string, string> = {
  pdf: "application/pdf",
  jpg: "image/jpeg",
  jpeg: "image/jpeg",
  png: "image/png",
  heic: "image/heic",
  heif: "image/heif",
  webp: "image/webp",
}

function resolveMimeType(file: File): string {
  if (file.type) return file.type
  const extension = file.name.split(".").pop()?.toLowerCase() ?? ""
  return EXTENSION_FALLBACK[extension] ?? ""
}

export function validateFile(file: File): FileValidationResult {
  const mimeType = resolveMimeType(file)

  if (!ACCEPTED_MIME_TYPES.includes(mimeType as (typeof ACCEPTED_MIME_TYPES)[number])) {
    return {
      valid: false,
      reason: {
        en: `${file.name} is not a supported file type. Use PDF, JPG, PNG or HEIC.`,
        hi: `${file.name} समर्थित फ़ाइल प्रकार नहीं है। PDF, JPG, PNG या HEIC उपयोग करें।`,
      },
    }
  }

  if (file.size > MAX_FILE_SIZE_BYTES) {
    const limit = formatFileSize(MAX_FILE_SIZE_BYTES)
    return {
      valid: false,
      reason: {
        en: `${file.name} is larger than ${limit}. Please upload a smaller file.`,
        hi: `${file.name} ${limit} से बड़ी है। कृपया छोटी फ़ाइल अपलोड करें।`,
      },
    }
  }

  return { valid: true }
}

export function createDocumentRecord(
  file: File,
  docType: DocumentTypeId,
): UploadedDocument {
  return {
    id: `doc-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
    name: file.name,
    size: file.size,
    mimeType: resolveMimeType(file),
    docType,
    stage: "queued",
    progress: 0,
    addedAt: new Date().toISOString(),
  }
}

/** Plausible extraction output so the review screen has something to show. */
function mockExtraction(document: UploadedDocument): DocumentExtraction {
  const label = DOCUMENT_TYPE_BY_ID[document.docType].label.en
  const fieldsByType: Record<DocumentTypeId, Array<{ label: string; value: string }>> = {
    prescription: [
      { label: "Prescriber", value: "Dr. R. Sharma" },
      { label: "Date", value: "05 Nov 2024" },
      { label: "Medications", value: "Amlodipine 5mg OD, Metformin 500mg BD" },
    ],
    lab: [
      { label: "Panel", value: "Complete Blood Count" },
      { label: "Haemoglobin", value: "11.2 g/dL (low)" },
      { label: "Collected", value: "15 Jan 2025" },
    ],
    discharge: [
      { label: "Facility", value: "Civil Hospital" },
      { label: "Admitted", value: "12 Feb 2024" },
      { label: "Discharged", value: "16 Feb 2024" },
    ],
    imaging: [
      { label: "Study", value: "Chest X-ray PA view" },
      { label: "Reported", value: "18 Mar 2025" },
    ],
    other: [{ label: "Document type", value: label }],
  }

  return {
    documentId: document.id,
    detectedType: document.docType,
    confidence: 0.92,
    fields: fieldsByType[document.docType],
  }
}

export interface ProcessOptions {
  /** Fired on every stage change so the UI can animate the pipeline. */
  onStage: (stage: ProcessingStage, progress: number) => void
  /** Milliseconds per stage. Kept short enough for a live demo. */
  stageDurationMs?: number
  signal?: AbortSignal
}

export const documentService = {
  validateFile,
  createDocumentRecord,

  /**
   * Runs one document through the pipeline.
   *
   * Today this is a timed simulation and the file never leaves the browser.
   * Against the backend it becomes a multipart POST plus a status poll (or an
   * SSE stream), reporting the same stages through `onStage`.
   */
  async process(
    document: UploadedDocument,
    file: File | undefined,
    options: ProcessOptions,
  ): Promise<DocumentExtraction> {
    const { onStage, stageDurationMs = 900, signal } = options

    if (!API_CONFIG.useMock && file) {
      const form = new FormData()
      form.append("file", file)
      form.append("docType", document.docType)
      onStage("uploading", 10)
      const extraction = await request<DocumentExtraction>("/documents", {
        method: "POST",
        body: form,
        signal,
      })
      onStage("complete", 100)
      return extraction
    }

    for (const [index, stage] of PIPELINE_STAGES.entries()) {
      if (signal?.aborted) throw new Error("aborted")
      const progress = Math.round(((index + 1) / PIPELINE_STAGES.length) * 100)
      onStage(stage, progress)
      if (stage !== "complete") await delay(stageDurationMs)
    }

    return mockExtraction(document)
  },

  /** Deletes a document. The mock has nothing server-side to clean up. */
  async remove(documentId: string): Promise<void> {
    if (!API_CONFIG.useMock) {
      await request<void>(`/documents/${documentId}`, { method: "DELETE" })
      return
    }
    await delay(60)
  },
}

File: backend/app/services/document_service.py
from __future__ import annotations

import uuid
from pathlib import Path

from fastapi import HTTPException, UploadFile
from sqlalchemy.orm import Session

from app.core.config import Settings
from app.models import Document, DocumentExtractionField

# Same allowlist as src/data/documentTypes.ts — kept in sync deliberately.
ACCEPTED_MIME_TYPES = {
    "application/pdf",
    "image/jpeg",
    "image/png",
    "image/heic",
    "image/heif",
    "image/webp",
}
DOCUMENT_TYPES = {"prescription", "lab", "discharge", "imaging", "other"}

# Same fixed fields as documentService.ts's `mockExtraction` — no OCR/AI runs
# here yet (Phase 8: "Do NOT implement OCR yet").
_MOCK_FIELDS: dict[str, list[dict]] = {
    "prescription": [
        {"label": "Prescriber", "value": "Dr. R. Sharma"},
        {"label": "Date", "value": "05 Nov 2024"},
        {"label": "Medications", "value": "Amlodipine 5mg OD, Metformin 500mg BD"},
    ],
    "lab": [
        {"label": "Panel", "value": "Complete Blood Count"},
        {"label": "Haemoglobin", "value": "11.2 g/dL (low)"},
        {"label": "Collected", "value": "15 Jan 2025"},
    ],
    "discharge": [
        {"label": "Facility", "value": "Civil Hospital"},
        {"label": "Admitted", "value": "12 Feb 2024"},
        {"label": "Discharged", "value": "16 Feb 2024"},
    ],
    "imaging": [
        {"label": "Study", "value": "Chest X-ray PA view"},
        {"label": "Reported", "value": "18 Mar 2025"},
    ],
    "other": [{"label": "Document type", "value": "Other Document"}],
}


def _validate(file: UploadFile, size: int, settings: Settings) -> None:
    if file.content_type not in ACCEPTED_MIME_TYPES:
        raise HTTPException(status_code=422, detail=f"Unsupported file type: {file.content_type}")
    if size > settings.max_upload_bytes:
        raise HTTPException(status_code=422, detail="File exceeds the maximum upload size")


def save_upload(
    db: Session,
    file: UploadFile,
    doc_type: str,
    document_session_id: str | None,
    settings: Settings,
) -> dict:
    if doc_type not in DOCUMENT_TYPES:
        raise HTTPException(status_code=422, detail=f"Unknown document type: {doc_type}")

    contents = file.file.read()
    _validate(file, len(contents), settings)

    document_id = f"doc-{uuid.uuid4().hex[:12]}"
    upload_dir = Path(settings.upload_dir)
    upload_dir.mkdir(parents=True, exist_ok=True)
    storage_path = upload_dir / document_id
    storage_path.write_bytes(contents)

    fields = _MOCK_FIELDS[doc_type]
    document = Document(
        id=document_id,
        patient_id=None,
        document_session_id=document_session_id,
        name=file.filename or document_id,
        size=len(contents),
        mime_type=file.content_type or "application/octet-stream",
        doc_type=doc_type,
        stage="complete",
        progress=100,
        storage_path=str(storage_path),
        detected_type=doc_type,
        confidence=0.92,
    )
    db.add(document)
    db.flush()
    for field in fields:
        db.add(DocumentExtractionField(document_id=document.id, label=field["label"], value=field["value"]))
    db.commit()

    return {
        "documentId": document.id,
        "detectedType": document.detected_type,
        "confidence": document.confidence,
        "fields": fields,
    }


def remove(db: Session, document_id: str, settings: Settings) -> bool:
    document = db.get(Document, document_id)
    if not document:
        return False
    if document.storage_path:
        Path(document.storage_path).unlink(missing_ok=True)
    db.delete(document)
    db.commit()
    return True
